server/webapp/WEB-INF/rails/webpack/views/pages/secret_configs/rules_widget.tsx

```diff
@@ -0,0 +1,189 @@
+    return [
+      {
+        id: "", text: "Select"
+      }
+      , {
+        id: "pipeline_group", text: "Pipeline Group"
+      }
+    ];
+  }
+
+  view(vnode: m.Vnode<AutoCompleteAttrs, this>): m.Children | void | null {
+    const removeRuleCallback = (ruleToBeRemoved: Stream<Rule>) => {
+      const index = vnode.attrs.rules().findIndex((r) => r === ruleToBeRemoved);
+      if (index !== -1) {
+        vnode.attrs.rules().splice(index, 1);
+      }
+    };
+    return <div data-test-id="rules-widget">
+      <h2>Rules </h2>
+      <FlashMessage type={MessageType.info}
+                    message="Configuring Rules is required to utilize this Secret Configuration. In absence of any rules, the secret configuration is denied access of any GoCD entities."/>
+      <div data-test-id="rules-table" className={styles.rulesTable}>
+        <Table headers={RulesWidget.headers()}
+               data={new RulesWidgetBody(vnode.attrs.rules,
+                                         vnode.attrs.resourceAutocompleteHelper,
+                                         removeRuleCallback).getData()}
+               draggable={true}
+               dragHandler={this.reArrange.bind(this, vnode.attrs.rules)}/>
+      </div>
+    </div>;
+  }
+
+  private reArrange(rule: Stream<Rules>, oldIndex: number, newIndex: number) {
+    const splicedRule = rule().splice(oldIndex, 1);
+    rule().splice(newIndex, 0, splicedRule[0]);
+    m.redraw();
+  }
+}
+
+class RulesWidgetBody {
+  private rules: Stream<Rules>;
+  private resourceAutocompleteHelper: Map<string, string[]>;
+  private removeRule: (ruleToBeRemoved: Stream<Rule>) => void;
+
+  constructor(rules: Stream<Rules>,
+              resourceAutocompleteHelper: Map<string, string[]>,
+              removeRule: (ruleToBeRemoved: Stream<Rule>) => void) {
+    this.rules                      = rules;
+    this.resourceAutocompleteHelper = resourceAutocompleteHelper;
+    this.removeRule                 = removeRule;
+  }
+
+  getData(): m.Child[][] {
+    return _.map(this.rules(), (rule) => {
+      const provider = stream(new ResourceSuggestionProvider(rule, this.resourceAutocompleteHelper));
+      return [<SelectField dataTestId="rule-directive"
+                           property={rule().directive}
+                           required={true}
+                           errorText={rule().errors().errorsForDisplay("directive")}>
+        <SelectFieldOptions selected={rule().directive()}
+                            items={RulesWidget.directives()}/>
+      </SelectField>,
+        <SelectField
+          dataTestId="rule-type"
+          property={rule().type}
+          required={true}
+          onchange={(e) => provider().update()}
+          errorText={rule().errors().errorsForDisplay("type")}>
+          <SelectFieldOptions selected={rule().type()}
+                              items={RulesWidget.types()}/>
+        </SelectField>,
+        <AutocompleteField
+          key={rule().type()}
```

Comment:
I don't think `key` here is meaningful in this context; it's not wrong, but it also doesn't affect the end result, as it really only matters on the outer reorderable set of children. I think you told me you attached a `key` attribute to the outer `<tr/>`, which seems like the right choice.

Reply:
With regards to the `<Table/>` component, I did see that `key` on the `<tr/>` was set to the `index` of the row. I may be mistaken as I'm basing this off a thought-experiment, but `index` may not be an effective key because it does not "follow" the data it is associated with; that means the `key` will always be reset to its current position rather than being associated with a specific piece of data, *I think*. `key` (to my understanding) is supposed to be the link between the underlying model and a given DOM node when elements are reorderable. Tying `key` to the `index` of a row might defeat the purpose.

I know I'm seeing some unintended behavior that *might* be related to this. For instance, when reordering rows while the autocomplete popup is open, the value of the `<input/>` reflects the reordering, but the popup selection does not, meaning mithril did not actually reorder the DOM, but just replaced the `<input/>` value, which sounds like it's related to a bad `key`, but that's just a guess.

Reply:
> I don't think `key` here is meaningful in this context; it's not wrong, but it also doesn't affect the end result, as it really only matters on the outer reorderable set of children. I think you told me you attached a `key` attribute to the outer `<tr/>`, which seems like the right choice.

This key attached over here is basically to force `mithril` to re-render the autocomplete text field when the provider updates.

Reply:
> With regards to the `<Table/>` component, I did see that `key` on the `<tr/>` was set to the `index` of the row. I may be mistaken as I'm basing this off a thought-experiment, but `index` may not be an effective key because it does not "follow" the data it is associated with; that means the `key` will always be reset to its current position rather than being associated with a specific piece of data, _I think_. `key` (to my understanding) is supposed to be the link between the underlying model and a given DOM node when elements are reorderable. Tying `key` to the `index` of a row might defeat the purpose.

I know that `index` isn't the right context for key, but there isn't any `identifier` associated with the data in the component as the input is `m.Child[][]`.